### PR TITLE
Fix Bezier intersection bug around the boundaries (issue #3381)

### LINKED ISSUE #3381: Issue with Bezier traits
Here is an old bug-report by @alextsui05

"Attached, I will leave a small program and two data files that show the problem.

Running the test with a.dat will demonstrate that we unexpectedly find an intersection point between the upper quarter circle curve and the lower quarter circle curve after being made x-monotone. We expect them to be disjoint because there is a tiny subcurve separating them.

b.dat replaces the lower quarter circle with a straight-line segment with the same endpoints. Running the test with this shows that no intersection point is found, as we expect.

I tracked the source of the intersection point to a call on line 2442 in the Arr_geometry_traits/Bezier_x_monotone_2.h file, in the _intersect method. The details are beyond me at the moment, but I guess a way to proceed would be to compare why a.dat and b.dat yield different results."

[argt_bezier.zip](https://github.com/CGAL/cgal/files/2454875/argt_bezier.zip)

### PR DESCRIPTION
**Pull request by theartful**
## Summary of Changes

Compute all roots of one of the resultants to do successful matching.

Two methods are used for computing the intersections:
1. de Casteljau subdivision.
2. Analytic solution by solving their equations.

In case de Casteljau fails, the analytic method is used. It's easy to reproduce the bug by disabling de Casteljau subdivision (changing line 2355 in `Bezier_x_monotone_2.h` for example to `if (false)`)
and  trying these curves:
https://www.desmos.com/calculator/f9efwzbb3k
In which, there are two roots, both out of range, but it'll match them as if they're one root resulting in an incorrect root which is in range.
## Release Management

* Affected package(s): 2D Arrangements
* Issue(s) solved (if any): fix #3381

* License and copyright ownership: license already used by CGAL

### CODE REVIEW

File: Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Bezier_cache.h [hunk 1]
```diff
@@ -606,21 +598,21 @@ bool _Bezier_cache<NtTraits>::_intersection_params
 
   for (k = degX_2; k >= 0; k--)
   {
-    coeff = nt_traits.get_coefficient (polyX_2, k) * normX_1;
+    coeff = nt_traits.get_coefficient (polyX_2, k);
```
Comment (theartful) [inline Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Bezier_cache.h]: Multiplying the polynomials with these huge normalizing factors really degraded the root finding algorithm. I can't see any reason why it is done. Shouldn't it be divided by it to get the original rational polynomial?

Removing these seemed to solve the problem. If they're necessary, I have another solution that works.

File: Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Bezier_cache.h [hunk 2]
```diff
@@ -170,6 +170,7 @@ template <class Nt_traits_> class _Bezier_cache
                                           tangency parameters. */
   Intersect_map     intersect_map;    /*! Maps curve pairs to their
                                           intersection parameter pairs. */
+  std::pair<Polynomial, Polynomial>     resultant_pair;
```
Comment (theartful) [inline Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Bezier_cache.h]: The resultant polynomial of the two bezier curves. Need to store them to do the boundary check.

File: Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Bezier_x_monotone_2.h [hunk 3]
```diff
@@ -1720,6 +1731,37 @@ bool _Bezier_x_monotone_2<RatKer, AlgKer, NtTrt, BndTrt>::_is_in_range
...
+  else
+  {
+    Originator_iterator  t_org = _pt.get_originator (_curve, _xid);
+    at_boundary = t_resultant.eval(t_org->point_bound().t_max);
+  }
+  at_t = at_t.abs();
+  at_boundary = at_boundary.abs();
+
+  if (CGAL::compare(at_t, at_boundary) == SMALLER)
+    return true;
```
Comment (theartful) [inline Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Bezier_x_monotone_2.h]: polynomial value check to see whether the boundary is closer to the real root than the calculated root

File: Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Bezier_x_monotone_2.h [hunk 4]
```diff
@@ -1720,6 +1731,37 @@ bool _Bezier_x_monotone_2<RatKer, AlgKer, NtTrt, BndTrt>::_is_in_range
...
+
+  Nt_traits nt_traits;
+  Polynomial derivative = nt_traits.derive(t_resultant);
+  Algebraic deriv_at_t = derivative.eval(t);
+  return (is_min && deriv_at_t.sign() < 0) || (!is_min && deriv_at_t.sign() > 0);
```
Comment (theartful) [inline Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Bezier_x_monotone_2.h]: first derivative check to see the direction on which the root resides

File: Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Bezier_x_monotone_2.h [hunk 5]
```diff
@@ -2639,6 +2681,17 @@ bool _Bezier_x_monotone_2<RatKer, AlgKer, NtTrt, BndTrt>::_intersect
     if (_is_in_range (iit->s, cache) &&
         cv._is_in_range (iit->t, cache))
     {
+      if (iit == inter_list.begin())
+      {
+        if (!_resultant_boundary_check(iit->t, cache, true))
```
Comment (theartful) [inline Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Bezier_x_monotone_2.h]: these boundary checks are only tested at the first and last calculated roots

**Comment by lrineau** [comment]
Hi @theartful, would that patch also apply to release branches `releases/CGAL-4.14-branch` and `releases/CGAL-5.0-branch`? If that is the case, the PR should be rebased onto `releases/CGAL-4.14-branch`.

**Comment by theartful** [comment]
Sorry about the hassle over such a small fix. Third time is the charm!
The problem was that roots were matched incorrectly. Specifically this case:
```
 (root1)      (root2)       (root3)
-------------=============================
                      curve2
=====================------------------------      
       curve1

===== : valid range
----- : invalid range
```
The previous method would calculate the roots in the valid range for each curve, which in this case would be:
```
curve1_roots: {root1_sval, root2_sval}
curve2_roots: {root2_tval, root3_tval}
```
And then would match t-s pair with the curve having the least roots to the one with the biggest number of roots. Clearly here `root2` and `root3` cannot be matched, but it would match them anyway resulting in erroneous results.
The solution I did is to calculate all the roots for one of the curves, then the matching will always be successful.

**Comment by lrineau** [comment]
@theartful Why do you frequently close and then reopen the PR?

**Comment by theartful** [comment]
Yeah, sorry about that. I found out that my previous attempt didn't solve
the root cause, and I thought that it will be a while till I find out
what's the problem. So I close it not to waste any reviewer's time.
Probably not the best move. Last time was just rebasing error.

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-98.shtml

**Comment by sloriot** [comment]
@theartful I've rebased your branch, please hard reset you local copy.

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-187.shtml

**Pull request merged**